packages/flutter/lib/src/widgets/interactive_viewer.dart

```diff
@@ -0,0 +1,1136 @@
+    final Quad correctedViewport = _transformViewport(correctedMatrix, _viewport);
+    final Offset offendingCorrectedDistance = _exceedsBy(boundariesAabbQuad, correctedViewport);
+    if (offendingCorrectedDistance == Offset.zero) {
+      return correctedMatrix;
+    }
+
+    // If the corrected translation doesn't fit in either direction, don't allow
+    // any translation at all. This happens when the viewport is larger than the
+    // entire boundary.
+    if (offendingCorrectedDistance.dx != 0.0 && offendingCorrectedDistance.dy != 0.0) {
+      return matrix;
+    }
+
+    // Otherwise, allow translation in only the direction that fits. This
+    // happens when the viewport is larger than the boundary in one direction.
+    final Offset unidirectionalCorrectedTotalTranslation = Offset(
+      offendingCorrectedDistance.dx == 0.0 ? correctedTotalTranslation.dx : 0.0,
+      offendingCorrectedDistance.dy == 0.0 ? correctedTotalTranslation.dy : 0.0,
+    );
+    return matrix.clone()..setTranslation(Vector3(
+      unidirectionalCorrectedTotalTranslation.dx,
+      unidirectionalCorrectedTotalTranslation.dy,
+      0.0,
+    ));
+  }
+
+  // Return a new matrix representing the given matrix after applying the given
+  // scale.
+  Matrix4 _matrixScale(Matrix4 matrix, double scale) {
+    if (widget.disableScale || scale == 1) {
+      return matrix;
+    }
+    assert(scale != 0.0);
+
+    // Don't allow a scale that results in an overall scale beyond min/max
+    // scale.
+    final double currentScale = widget.transformationController.value.getMaxScaleOnAxis();
+    final double totalScale = currentScale * scale;
+    final double clampedTotalScale = totalScale.clamp(
+      widget.minScale,
+      widget.maxScale,
+    ) as double;
+    final double clampedScale = clampedTotalScale / currentScale;
+    final Matrix4 nextMatrix = matrix.clone()..scale(clampedScale);
+
+    // Ensure that the scale cannot make the child so big that it can't fit
+    // inside the boundaries (in either direction).
+    final double minScale = math.max(
+      _viewport.width / _boundaryRect.width,
+      _viewport.height / _boundaryRect.height,
+    );
+    if (clampedTotalScale < minScale) {
+      final double minCurrentScale = minScale / currentScale;
+      return matrix.clone()..scale(minCurrentScale);
+    }
+
+    return nextMatrix;
+  }
+
+  // Return a new matrix representing the given matrix after applying the given
+  // rotation.
+  Matrix4 _matrixRotate(Matrix4 matrix, double rotation, Offset focalPoint) {
+    if (widget.disableRotation || rotation == 0) {
+      return matrix;
+    }
+    final Offset focalPointScene = widget.transformationController.toScene(
+      focalPoint,
+    );
+    return matrix
+      ..translate(focalPointScene.dx, focalPointScene.dy)
+      ..rotateZ(-rotation)
+      ..translate(-focalPointScene.dx, -focalPointScene.dy);
+  }
+
+  // Handle the start of a gesture of _GestureType. All of translation, scale,
```

Comment:
Does this doc need to be updated? We seem to be just handling a scale gesture here. Or maybe you just mean "_gestureType" (here and elsewhere).

Reply:
I'm using the onScale series of events from GestureDetector, but then I interpret the gesture as being _GestureType.scale, pan, or rotate.  I'll clarify the comments.